### PR TITLE
feat(abstract-eth): wrw recovery for dkls wallet

### PR DESCRIPTION
**Pull request by dpkjnr**
Ticket: WP-1885

### CODE REVIEW

**Review by gianchandania — DISMISSED** [review]
lgtm

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts [hunk 1]
```diff
@@ -1926,35 +1930,100 @@ export abstract class AbstractEthLikeNewCoins extends AbstractEthLikeCoin {
...
+    );
+    const sigParts = signatureString.split(':');
+
+    return {
+      recid: parseInt(sigParts[0], 10),
+      r: sigParts[1],
+      s: sigParts[2],
+      y: sigParts[3],
+    };
+  }
```
Comment (zahin-mohammad) [inline modules/abstract-eth/src/abstractEthLikeNewCoins.ts:2006]: Dont we have a re-usable function from gg18 tss for this?

**Review by zahin-mohammad — DISMISSED** [review]
lgtm, have we tested this with an actual dkls wallet yet?

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts (hunk 1 above)
Comment (dpkjnr) [reply modules/abstract-eth/src/abstractEthLikeNewCoins.ts:2009]: I didn't find it in the utils

**Comment by dpkjnr** [comment]
tested recovery using locally linking sdk modules to WRW. Used the following data

baseAddress
0xb6c3ae810db93d67f28476bf74f497a2bfbdbffa

ReducedUserShare
{"iv":"uWPkYuoW/fN8We95LNp9+Q==","v":1,"iter":1000,"ks":128,"ts":64,"mode":"ccm","adata":"","cipher":"aes","salt":"astk9Wft9Gw=","ct":"Q5Vff3XZSUEpivhWnImyvjF8Dv6EeawLgAjiGWhD61AWU1WdhpKNk8nPxrgP0swRk0lTdHuesOETRKO+C1m86ZPbFsWqjasPa6LB1crWvc1dLhsD9QKZK3U86OoFfXLqsCICyZk0fbB0CaoTEfEqVsrQgcio6HVvpqlU+G+KRBjr697apNP+LJF5sqUbRkUFv7+A1TOOkWK/qcmGu5xPSGJSc1WM6II8DqaLREvbdWxyZH+OwZS+IFOYdU8566WbX+vDPtRD5p4xzD8XbVWnMaZZvGLjHjz2fdVAvDyE0T7bdGiG2Gho+l/uM6f3Dkcg3P2SaXgReEmwz4CQR5+3ILFmwvQ6GlFZ2Uhdu4gQZLA/1lAVDlZV7RruXPj9B4Xu8TPPo4MbXCBTI44zamR+SIsAtRFHRB7t/5glK0IkwGHS+y0yW6TFXoMhLVX0Y/zWD5SoHxjDzGRmTexzDkaPUeh+QIlBvyrS4ILUBefEYYrxX/ua3Zckwuv4E5F/FzpYunYqE2J06tLo6usHU6M8Q+XqyPYSyY9LvOC3utZyTcDJVeuRoxSeDokCIF7b1e3ZfDHow3hgarC0gnWjmWprRMPZGgqcabA0fCSEeY9JAvrYw1kGcHZrBocwfdURncahA4maANDxm7xEu0sEt9bjaP+diSUEE3OieC6D5tu4gviaxn78mfx/ZtkN+me6L5eT2rN+JSZozBzPWvtJ0BRpHufHTuLYTtjljW0G+i3bHgnv+ltiVFGjA7Vw3OvY0WeEpEyoaft0JuFbWjuBWelMu2vXbaopLaEDdgXz5czqOzV5MzpuK4gIMuiWVk3//fYrQePuKs6IN5oqXv27nRSDY4XdoPei/baDpqaK/x3vPYxaP62AXT3wL1/f6WWxDLxPHniCo0btrilBjPJxdYidGPCGY7PgiHuAX5cdocXsu/WlVMnrcWdHqsOsPrxpkHpaEkA3R10Ic/TDDOMoBPIPSVaN3aZBvSJHD5TD0c2isqF3j0Qx8OqFWdHdXjGRTKp9vjWtdTEsjmT8t6a9vCAyrzv1GJJKWbvocW3Nntx0R8W/f6xO+9or4cYkS2yTpy1XuzVWCeB5aBmFbFlW"}

ReducedBackupShare
{"iv":"dzgY7TreL3msbs45x8gjuA==","v":1,"iter":1000,"ks":128,"ts":64,"mode":"ccm","adata":"","cipher":"aes","salt":"astk9Wft9Gw=","ct":"dC1lRj1dlBF2bD08PHaclDmaZ2Lz3w+76YD7kZdHOsNJrN3WF9NFZjCuxl2DGmuLHNpgnxutdgnDZRBjo4++mslXKXH6aSrGOnaozUfWb/nhOHDVSkYTCLaF3aBd18LzEroVuKy0dGd+pLzeCxuwhTNCzAtUd2KNe65+tF9txHmreUTlYloY4T/YsBtG0DbSGkt+Nyl8oF2BSrmjaD/iFgaQ8vba69mBnRrjcI91Ng6vYdcNOxdrncx4f2+hGPAEyNu/tFWEwTIcGzuCt3lP70Sc1gm/AsyBzw4/2YTdxo1N/dejGhjS4VPHWm6TRHk57plCftb8Zoek9rFm38SrKyYAqmmeERqBVKbW1RrE/BfsaipWDuYTHLbLmPRULbX1wjYa1As+5NwDQVCUSrtUwH4aqmmzZ/18mDXGykJd4fjsU/UCTd49f+/RqOg73ioioILmCcCGKLAfb7I2IMExkAOd+PdOJh3zVWsgLSAbe1DjIIoK/cFS+rA2qK6QwS5Xdmpl22w8N++EXBgSb4v9HHZPRiPNoiJ31J2cT7cIQiYJkAdTDVnXCM864X3K7IjljJl1zyYwKhUXNauQjWoOIkVb6Hc8Og8V0P8XJunwIDDkK1JG+SFOEPXRZKSE2xhc7oRJ9Or6D0uwVNKFh4QLbyThlKBFSwGtNZuHm4rzCsRocjwurca5c445XzsOAk6mM+0UQ5Fo2jJTFDtXV4DzYhsqQt7btUq3oaiHsGGPtwNRr1tDTjIWlpLdw/c4Wm6kdFhXrgCfjYHytk4zyr01K+PacNhnr4Ls4G6WOz84Dt036vEdhEsvBEYlzIcv6io1z5qY6iYHY2DmDOBW12YqNg71Bvzz4A0qncZLwo356kI9sq4QesQIZ+EScG83HNGS0arFpkzVJTUHuho60qAzfl3TLUsXWnZKoSrRaUHM+R9BL0sZj+zZjcVzsiwHQ833q7gORWjBwrRykZzYNeS3MuvnzS8PDNUL3aRxiPCx5nvSD232Q+UJZkQ6AFZYzdGRtLpbYAZghYl4ztuZRTRWUZKmmlqWJu7yMrWDsOzoluE6IfYA6/hDS/0tdCJmHAmkgRo0jU4raMQ="}


Recovery transaction:
https://holesky.etherscan.io/tx/0x34e83cecc9a44568c6b42b12d0d2fd014171eae7897759790f47de42a71d7718

**Comment by dpkjnr** [comment]
> lgtm, have we tested this with an actual dkls wallet yet?

I have tested using test data generated in the testcases. Is keycard with reduced keyshare working now or we need to update SDK version in bitgo-ui?

**Comment by dpkjnr** [comment]
Printed reduced keyshares while creating wallet via SDK script and tested recovery for wallet 664b83a3935a72fd07bf16b1048ff01a, assumed that reducedKeyshares will be printed on the keycard

recovery txn: https://holesky.etherscan.io/tx/0x66c078970dd103f697bceb6f3d5241a93ddecf24133afc92286b032f2dd994d1

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts [hunk 2]
```diff
@@ -70,6 +70,9 @@ import {
   TransactionBuilder,
   TransferBuilder,
 } from './lib';
+import { decode } from 'cbor-x';
+import { isRight } from 'fp-ts/Either';
+import { Buffer } from 'buffer';
```
Comment (alebusse) [inline modules/abstract-eth/src/abstractEthLikeNewCoins.ts:75]:
```suggestion
```
I wonder if we could move all this gg18 <> mpcv2 logic to the mpc lib as utils so we dont need to install `cbor-x` and `fp-ts/Either`

Comment (dpkjnr) [reply modules/abstract-eth/src/abstractEthLikeNewCoins.ts]: created utility function to decode ReducedKeyPair and removed direct cbor-x and fp-ts dependency

**Review by gianchandania — APPROVED** [review]

**Review by ayush-9 — APPROVED** [review]

**Comment by zahin-mohammad** [comment]
> > lgtm, have we tested this with an actual dkls wallet yet?
> 
> I have tested using test data generated in the testcases. Is keycard with reduced keyshare working now or we need to update SDK version in bitgo-ui?

Yes reduced keycard should be working now (at least in the SDK, but maybe not in the UI).

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts (hunk 2 above)
Comment (zahin-mohammad) [reply modules/abstract-eth/src/abstractEthLikeNewCoins.ts]: the sdk-lib-mpc should just be the mpc primitives, lets not couple that to bitgo specific encodings.

File: modules/sdk-lib-mpc/src/tss/ecdsa-dkls/util.ts [hunk 3]
```diff
@@ -79,3 +82,174 @@ export function verifyAndConvertDklsSignature(
...
+  const backup = new Dkg(3, 2, 1, retrofitDataB);
+  const bitgo = new Dkg(3, 2, 2, retrofitDataC);
+  const userRound1Message = await user.initDkg();
+  const backupRound1Message = await backup.initDkg();
+  const bitgoRound1Message = await bitgo.initDkg();
+  const userRound2Messages = user.handleIncomingMessages({
+    p2pMessages: [],
+    broadcastMessages: [bitgoRound1Message, backupRound1Message],
+  });
+  const backupRound2Messages = backup.handleIncomingMessages({
```
Comment (zahin-mohammad) [inline modules/sdk-lib-mpc/src/tss/ecdsa-dkls/util.ts:155]: nit: sdk-lib-mpc should be agnostic to bitgo modeling, so we should remove the user/backup/bitgo naming scheme where possible.

File: modules/sdk-lib-mpc/src/tss/ecdsa-dkls/util.ts [hunk 4]
```diff
@@ -79,3 +82,174 @@ export function verifyAndConvertDklsSignature(
     dklsSignature.S
   ).toString('hex')}${delimeter}${truePub}`;
 }
+
+export async function executeTillRound(
+  round: number,
+  party1Dsg: Dsg,
+  party2Dsg: Dsg
+): Promise<DeserializedMessages[] | DeserializedDklsSignature> {
```
Comment (zahin-mohammad) [inline modules/sdk-lib-mpc/src/tss/ecdsa-dkls/util.ts:90]:
```suggestion
export async function executeTillRound(
  party1Dsg: Dsg,
  party2Dsg: Dsg,
  untilRound = 5,
): Promise<DeserializedMessages[] | DeserializedDklsSignature> {
```

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts [hunk 5]
```diff
@@ -1926,35 +1930,130 @@ export abstract class AbstractEthLikeNewCoins extends AbstractEthLikeCoin {
...
+    backupKeyShare: Buffer,
+    commonKeyChain: string
+  ) {
+    const messageHash = tx.getMessageToSign(true);
+    const userDsg = new DklsDsg.Dsg(userKeyShare, 0, 'm', messageHash);
+    const backupDsg = new DklsDsg.Dsg(backupKeyShare, 1, 'm', messageHash);
+
+    const signatureString = DklsUtils.verifyAndConvertDklsSignature(
+      messageHash,
+      (await DklsUtils.executeTillRound(5, userDsg, backupDsg)) as DklsTypes.DeserializedDklsSignature,
```
Comment (zahin-mohammad) [inline modules/abstract-eth/src/abstractEthLikeNewCoins.ts:1992]: Why do we need a cast here?

**Review by zahin-mohammad — APPROVED** [review]

**Pull request merged**